src/wp-includes/class-wp-block.php

```diff
@@ -191,6 +191,200 @@ public function __get( $name ) {
 		return null;
 	}
 
+	/**
+	 * Processes the block bindings in block's attributes.
+	 *
+	 * A block might contain bindings in its attributes. Bindings are mappings
+	 * between an attribute of the block and a source. A "source" is a function
+	 * registered with `register_block_bindings_source()` that defines how to
+	 * retrieve a value from outside the block, e.g. from post meta.
+	 *
+	 * This function will process those bindings and replace the HTML with the value of the binding.
+	 * The value is retrieved from the source of the binding.
+	 *
+	 * ### Example
+	 *
+	 * The "bindings" property for an Image block might look like this:
+	 *
+	 * ```json
+	 * {
+	 *   "metadata": {
+	 *     "bindings": {
+	 *       "title": {
+	 *         "source": "post_meta",
+	 *         "args": { "key": "text_custom_field" }
+	 *       },
+	 *       "url": {
+	 *         "source": "post_meta",
+	 *         "args": { "key": "url_custom_field" }
+	 *       }
+	 *     }
+	 *   }
+	 * }
+	 * ```
+	 *
+	 * The above example will replace the `title` and `url` attributes of the Image
+	 * block with the values of the `text_custom_field` and `url_custom_field` post meta.
+	 *
+	 * @access private
+	 * @since 6.5.0
+	 *
+	 * @param string   $block_content Block content.
+	 * @param array    $block The full block, including name and attributes.
+	 */
+	private function process_block_bindings( $block_content ) {
+		$block = $this->parsed_block;
+
+		// Allowed blocks that support block bindings.
+		// TODO: Look for a mechanism to opt-in for this. Maybe adding a property to block attributes?
+		$allowed_blocks = array(
+			'core/paragraph' => array( 'content' ),
+			'core/heading'   => array( 'content' ),
+			'core/image'     => array( 'url', 'title', 'alt' ),
+			'core/button'    => array( 'url', 'text' ),
+		);
+
+		// If the block doesn't have the bindings property or isn't one of the allowed block types, return.
+		if ( ! isset( $block['attrs']['metadata']['bindings'] ) || ! isset( $allowed_blocks[ $this->name ] ) ) {
+			return $block_content;
+		}
+
+		$block_bindings_sources = get_all_registered_block_bindings_sources();
+		$modified_block_content = $block_content;
+		foreach ( $block['attrs']['metadata']['bindings'] as $binding_attribute => $binding_source ) {
```

Comment:
I think it would be better to check in advance that `$block['attrs']['metadata']['bindings']` is an array. For example, the following markup outputs a warning error.

```html
<!-- wp:paragraph {"metadata":{"bindings":"string"}} -->
<p>Hello</p>
<!-- /wp:paragraph -->
```

Reply:
Isn't that doing it wrong though? I mean I expect that you can have similar issues if you provide invalid values for a lot of block attributes no?

Reply:
Indeed. Although I have not investigated all block supports, if extenders intentionally rewrite an attribute to an invalid value, the error may occur with other block supports as well.

Reply:
It's true, but it's probably slightly better to just output the `$block_content` (i.e. fail silently) instead of showing an error in this case. I addressed it in 604f7a2cb3767163c7d12a4f571f76b7a47d20a6